ext/php8/handlers_curl.c

```diff
@@ -79,54 +76,44 @@ static void dd_ch_duplicate_headers(zval *ch_orig, zval *ch_new) {
     }
 }
 
-static void dd_init_headers_arg(zval *arg, zval *ch) {
-    HashTable *retval;
-    ALLOC_HASHTABLE(retval);
-    HashTable *headers = NULL;
-
-    if (dd_headers) {
-        headers = zend_hash_index_find_ptr(dd_headers, Z_OBJ_HANDLE_P(ch));
-        if (headers) {
-            size_t headers_count = zend_hash_num_elements(headers);
-            zend_hash_init(retval, headers_count, NULL, ZVAL_PTR_DTOR, 0);
-            zend_hash_copy(retval, headers, (copy_ctor_func_t)zval_add_ref);
-        }
+static void dd_inject_distributed_tracing_headers(zval *ch) {
+    zval headers;
+    zend_array *dd_header_array;
+    if (dd_headers && (dd_header_array = zend_hash_index_find_ptr(dd_headers, Z_OBJ_HANDLE_P(ch)))) {
+        ZVAL_ARR(&headers, zend_array_dup(dd_header_array));
+    } else {
+        array_init(&headers);
     }
 
-    if (!headers) {
-        zend_hash_init(retval, 0, NULL, ZVAL_PTR_DTOR, 0);
+    int sampling_priority;
+    if (ddtrace_fetch_prioritySampling_from_root(&sampling_priority)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-sampling-priority: %d", sampling_priority));
+    }
+    if (DDTRACE_G(trace_id)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-trace-id: %" PRIu64, (DDTRACE_G(trace_id))));
+    }
+    if (DDTRACE_G(span_ids_top)) {
+        add_next_index_str(&headers,
+                           zend_strpprintf(0, "x-datadog-parent-id: %" PRIu64, (DDTRACE_G(span_ids_top)->id)));
+    }
+    if (DDTRACE_G(dd_origin)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-origin: %s", ZSTR_VAL(DDTRACE_G(dd_origin))));
     }
-    ZVAL_ARR(arg, retval);
-}
 
-static void dd_free_headers_arg(zval *arg) { zend_array_destroy(Z_ARRVAL_P(arg)); }
+    zend_function *setopt_fn = zend_hash_str_find_ptr(EG(function_table), ZEND_STRL("curl_setopt"));
```

Comment:
Would you see any value in obtaining this once on startup vs every function call?

Reply:
I have a tiny bit of experience here. It would need to be looked up on every request, not just once on startup. Remember the sigsegv when my curl cache wasn't cleared properly? It would be similar to that. _Most_ of the time it won't move address, but sometimes it will.

Reply:
> Remember the sigsegv when my curl cache wasn't cleared properly?

🤔 I'm not really sure which issue you're referring to. I remember one issue came about because the request-scoped global that stored the function call info cache wasn't being initialized on RINIT, but I'm assuming you're referring to a different issue?

> Most of the time it won't move address, but sometimes it will.

If it is read out of the compiler globals on startup, what would cause the address to change?

If we can't look it up at startup, perhaps we could look it up once per request? I don't think it's too big of an issue either way. Just thinking of how we can reduce the number of hash table lookups. :)

Reply:
@morrisonlevi The zval containing the Z_PTR might be moved, yes (due to function table resizing), but the container zend_function * pointer should never be different from MINIT, or am I missing something here?

Reply:
I can't remember off the top of my head exactly what was being stored and caused a sigsegv. I do remember that it went through all tests and didn't fail until it reached customers.
I tried finding the PR but didn't find it right away; may have more luck looking through JIRA or release notes.

Reply:
I cannot imagine it being an issue, but in that case let's play safe, the overhead of fetching the function from the function table is vanishingly tiny compared to the actual curl request.